# Homework 6

Yihao Zhong

## Problem 1

### 1.

Model interpretability and explainability are important because they enable trust, transparency, and account for robustness for downstream users. Interpretable models allow downstream users to understand how predictions are made, facilitating debugging, improvement, and also compliance with regulations (equality, non-descriminatic etc). Explainability helps detect biases, uncover insights, and enhance user acceptance. Without interpretability, the decision-making process remains a "black box," hindering trust and limiting the model's practical applicability in domains that require clear explanations, such as healthcare and finance.

### 2.

In [ ]:
!pip install autofeat

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from autofeat import FeatureSelector, AutoFeatRegressor
from sklearn import datasets
from sklearn.linear_model import LinearRegression

In [ ]:
# Load the diabetes dataset and get the featues and target
X, y = datasets.load_diabetes(return_X_y=True)

print("Original number of features: ", X.shape[1])

Original number of features:  10


In [ ]:
fsel = FeatureSelector(verbose=1)

# Fit it on the train dataset
df = fsel.fit_transform(X, y)
print("New number of features: ", df.shape[1])


[featsel] Scaling data...

2024-04-27 13:37:15,434 INFO: [featsel] Feature selection run 1/5


done.


2024-04-27 13:37:16,154 INFO: [featsel] Feature selection run 2/5
2024-04-27 13:37:16,188 INFO: [featsel] Feature selection run 3/5
2024-04-27 13:37:16,220 INFO: [featsel] Feature selection run 4/5
2024-04-27 13:37:16,251 INFO: [featsel] Feature selection run 5/5
2024-04-27 13:37:16,284 INFO: [featsel] 7 features after 5 feature selection runs
/Users/zhongyihao/anaconda3/envs/trading/lib/python3.12/site-packages/autofeat/featsel.py:270: FutureWarning: Series.ravel is deprecated. The underlying array is already 1D, so ravel is not necessary.  Use `to_numpy()` for conversion to a numpy array instead.
  if np.max(np.abs(correlations[c].ravel()[:i])) < 0.9:
2024-04-27 13:37:16,286 INFO: [featsel] 7 features after correlation filtering
2024-04-27 13:37:16,295 INFO: [featsel] 6 features after noise filtering


New number of features:  6


Answer:

There are 10 original features, and after features selection there are 6 features, the discarded features is 4.

### 3.

In [ ]:
# Perform a train-test split on your dataset. Select a regression model from skLearn and fit it to the
# training dataset. What is the R2 score on the training and test set?

# Perform a train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Fit the model
rreg_model = LinearRegression()
result_logistic = rreg_model.fit(X_train, y_train)

print("R2 score on the training set: ", result_logistic.score(X_train, y_train))
print("R2 score on the test set: ", result_logistic.score(X_test, y_test))


R2 score on the training set:  0.5279193863361498
R2 score on the test set:  0.45260276297191915


We perform a linear regression model. The $R^2$ for training set is 0.53, and the $R^2$ for testing set is 0.45.

### 4.

In [ ]:
# Instantiate an AutoFeatRegressor with 3 steps of feature engineering
afreg = AutoFeatRegressor(verbose=1, feateng_steps=3)

# Fit it on the train dataset
X_train_afreg = afreg.fit_transform(X_train, y_train)
X_test_afreg = afreg.transform(X_test)

result_afreg = rreg_model.fit(X_train_afreg, y_train)

print("## Final Train R^2: %.4f" % result_afreg.score(X_train_afreg, y_train))
print("## Final Test R^2: %.4f" %  result_afreg.score(X_test_afreg, y_test))


2024-04-27 13:37:30,791 INFO: [AutoFeat] The 3 step feature engineering process could generate up to 60445 features.
2024-04-27 13:37:30,792 INFO: [AutoFeat] With 353 data points this new feature matrix would use about 0.09 gb of space.
2024-04-27 13:37:30,796 INFO: [feateng] Step 1: transformation of original features


2024-04-27 13:37:31,819 INFO: [feateng] Generated 45 transformed features from 10 original features - done.
2024-04-27 13:37:31,824 INFO: [feateng] Step 2: first combination of features


2024-04-27 13:37:33,274 INFO: [feateng] Generated 5789 feature combinations from 1485 original feature tuples - done.
2024-04-27 13:37:33,279 INFO: [feateng] Step 3: transformation of new features


/Users/zhongyihao/anaconda3/envs/trading/lib/python3.12/site-packages/numpy/core/_methods.py:176: RuntimeWarning: overflow encountered in multiply
  x = um.multiply(x, x, out=x)
/Users/zhongyihao/anaconda3/envs/trading/lib/python3.12/site-packages/numpy/core/_methods.py:187: RuntimeWarning: overflow encountered in reduce
  ret = umr_sum(x, axis, dtype, out, keepdims=keepdims, where=where)


2024-04-27 13:37:43,013 INFO: [feateng] Generated 24430 transformed features from 5789 original features - done.
2024-04-27 13:37:43,053 INFO: [feateng] Generated altogether 32266 new features in 3 steps
2024-04-27 13:37:43,053 INFO: [feateng] Removing correlated features, as well as additions at the highest level


2024-04-27 13:37:43,270 INFO: [feateng] Generated a total of 14871 additional features


[featsel] Scaling data...

2024-04-27 13:37:44,076 INFO: [featsel] Feature selection run 1/5


done.


2024-04-27 13:38:28,567 INFO: [featsel] Feature selection run 2/5
2024-04-27 13:39:09,195 INFO: [featsel] Feature selection run 3/5
2024-04-27 13:39:38,424 INFO: [featsel] Feature selection run 4/5
2024-04-27 13:40:33,087 INFO: [featsel] Feature selection run 5/5
2024-04-27 13:41:01,473 INFO: [featsel] 36 features after 5 feature selection runs
/Users/zhongyihao/anaconda3/envs/trading/lib/python3.12/site-packages/autofeat/featsel.py:270: FutureWarning: Series.ravel is deprecated. The underlying array is already 1D, so ravel is not necessary.  Use `to_numpy()` for conversion to a numpy array instead.
  if np.max(np.abs(correlations[c].ravel()[:i])) < 0.9:
2024-04-27 13:41:01,487 INFO: [featsel] 31 features after correlation filtering
2024-04-27 13:41:01,685 INFO: [featsel] 11 features after noise filtering
2024-04-27 13:41:01,705 INFO: [AutoFeat] Computing 11 new features.


2024-04-27 13:41:02,327 INFO: [AutoFeat]    11/   11 new features ...done.
2024-04-27 13:41:02,335 INFO: [AutoFeat] Final dataframe with 21 feature columns (11 new).
2024-04-27 13:41:02,335 INFO: [AutoFeat] Training final regression model.
2024-04-27 13:41:02,344 INFO: [AutoFeat] Trained model: largest coefficients:
2024-04-27 13:41:02,345 INFO: -70.30327923185092
2024-04-27 13:41:02,345 INFO: 781094.306607 * x000**3*x001
2024-04-27 13:41:02,346 INFO: -3983.086911 * exp(x006)*Abs(x001)
2024-04-27 13:41:02,346 INFO: 362.715436 * exp(x002)*exp(x003)
2024-04-27 13:41:02,347 INFO: 332.458747 * x000**9/x009**3
2024-04-27 13:41:02,347 INFO: 215.883125 * Abs(x002 + Abs(x009))
2024-04-27 13:41:02,347 INFO: 37.282823 * exp(x002)*exp(x008)
2024-04-27 13:41:02,348 INFO: 24.680193 * Abs(x008)/x008
2024-04-27 13:41:02,348 INFO: -4.397543 * x004**2*Abs(1/x003)
2024-04-27 13:41:02,348 INFO: -0.854832 * x006/Abs(x002)
2024-04-27 13:41:02,349 INFO: 0.037580 * 1/(x003**2 - x008)
2024-04-27 13:41:02,349 

## Final Train R^2: 0.6224 features
## Final Test R^2: 0.5172


In [ ]:
print("New Features: ", set(X_train_afreg.columns))
print("Original Features: ", datasets.load_diabetes().feature_names)

New Features:  {'x001', 'x004**2*Abs(1/x003)', 'x003', 'x006/Abs(x002)', 'Abs(x008)/x008', 'x004', 'x002', 'exp(x006)*Abs(x001)', 'x000**9/x009**3', 'exp(x002)*exp(x003)', 'Abs(x002 + Abs(x009))', 'x008', 'x009', '1/(x008 - Abs(x009))', 'x005', 'x000', 'x000**3*x001', 'x007', '1/(x003**2 - x008)', 'exp(x002)*exp(x008)', 'x006'}
Original Features:  ['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6']


The $R^2$ score on the train data is 0.62, increased by 17%. The $R^2$ score on the test data is 0.45, increased by 14.3%. The performance improve after we use the Autofeat, which introduce more features and hence more complexity and non-linearity to the regression model.

Five new features are 'x004**2*Abs(1/x003)', 'x006/Abs(x002)', 'Abs(x008)/x008', 'exp(x006)*Abs(x001)', 'exp(x002)*exp(x003)'.

## Problem 2

### 1.

In [2]:
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from sklearn.neighbors import KNeighborsClassifier
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import MinMaxScaler
from sklearn import datasets, neighbors, linear_model
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from scipy.stats import uniform

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.datasets import mnist

In [3]:
X_digits, y_digits = datasets.load_digits(return_X_y=True)
X_digits = X_digits / X_digits.max()

X_train, X_test, y_train, y_test = train_test_split(X_digits, y_digits,
                                                    test_size=0.2, random_state=42)

In [33]:
pip install "pydantic<2"

In [4]:
pip install ray[tune]

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.4/65.4 MB 23.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 101.7/101.7 kB 13.5 MB/s eta 0:00:00


In [5]:
import ray
from ray import tune
from ray.tune.integration.keras import TuneReportCallback

In [13]:
pip install "pydantic<2"

In [18]:
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader, TensorDataset
from torchvision import transforms
from torchvision.datasets import MNIST
from ray.tune.integration.keras import TuneReportCallback
from ray import train
from ray.air import session

# Define the Lenet Model
class LeNet(nn.Module):
    def __init__(self, num_filters, dropout_prob):
        super(LeNet, self).__init__()
        self.conv1 = nn.Conv2d(1, num_filters, kernel_size=5)
        self.conv2 = nn.Conv2d(num_filters, 50, kernel_size=5)
        self.dropout = nn.Dropout(dropout_prob)
        self.fc1 = nn.Linear(800, 500)
        self.fc2 = nn.Linear(500, 10)

    def forward(self, x):
        x = torch.relu(self.conv1(x))
        x = torch.max_pool2d(x, 2)
        x = torch.relu(self.conv2(x))
        x = torch.max_pool2d(x, 2)
        x = x.view(-1, 800)
        x = torch.relu(self.fc1(x))
        x = self.dropout(x)
        x = self.fc2(x)
        return x

def train_mnist(config):
    num_epochs = 5
    num_classes = 10
    #device = "cuda" if torch.cuda.is_available() else "cpu"

    # Load and transform the MNIST dataset
    transform = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,))
    ])

    train_dataset = MNIST(root='./data', train=True, transform=transform, download=True)
    test_dataset = MNIST(root='./data', train=False, transform=transform, download=True)

    train_loader = DataLoader(dataset=train_dataset, batch_size=int(config["batch_size"]), shuffle=True)
    test_loader = DataLoader(dataset=test_dataset, batch_size=int(config["batch_size"]), shuffle=False)
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    # Initialize the model
    model = LeNet(num_filters=int(config["filters"]), dropout_prob=config["dropout"]).to(device)
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.Adam(model.parameters(), lr=config["lr"])

    # Training loop
    for epoch in range(num_epochs):
        model.train()
        for batch_idx, (data, target) in enumerate(train_loader):
            data, target = data.to(device), target.to(device)
            optimizer.zero_grad()
            output = model(data)
            loss = criterion(output, target)
            loss.backward()
            optimizer.step()

        # Evaluation
        model.eval()
        correct = 0
        total = 0
        with torch.no_grad():
            for data, target in test_loader:
                data, target = data.to(device), target.to(device)
                outputs = model(data)
                _, predicted = torch.max(outputs.data, 1)
                total += target.size(0)
                correct += (predicted == target).sum().item()

        accuracy = correct / total
        print(f"Epoch {epoch+1}/{num_epochs}, Accuracy: {accuracy}")

        # Tune report callback
        train.report(metrics={"accuracy": accuracy})


In [19]:
ray.shutdown()
ray.init(log_to_driver=False)

/usr/lib/python3.10/subprocess.py:1796: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  self.pid = _posixsubprocess.fork_exec(
2024-05-02 01:20:55,226	INFO worker.py:1749 -- Started a local Ray instance.


RayContext(dashboard_url='', python_version='3.10.12', ray_version='2.20.0', ray_commit='5708e75978413e46c703e44f43fd89769f3c148b')

In [10]:
import ray
from ray import tune
from ray.tune.schedulers import HyperBandScheduler
from ray.tune.search.bayesopt import BayesOptSearch


In [16]:

import time
# Define the search space for hyperparameters
search_space = {
    "filters": tune.choice([64, 128, 256]),
    "lr": tune.choice([0.001, 0.01, 0.1]),
    "batch_size": tune.choice([64, 128, 256]),
    "dropout": tune.uniform(0, 1),
}

# Set up Ray Tune


start_time = time.time()
# Grid search setup
grid_search_analysis = tune.run(
    train_mnist,
    num_samples=6,
    config=search_space,
    scheduler=None,  # No scheduler needed for grid search
    verbose=1,
     resources_per_trial={"cpu": 12, "gpu": 1}
)
end_time = time.time()
print("Time taken for grid search:", end_time - start_time)



# Printing best hyperparameters
print("Best hyperparameters from Grid Search:", grid_search_analysis.get_best_config(metric="accuracy"))


2024-05-01 00:32:30,743	INFO tune.py:762 -- Total run time: 1019.62 seconds (1019.49 seconds for the tuning loop).


Time taken for grid search: 1019.6484160423279


ValueError: No `mode` has been passed and  `default_mode` has not been set. Please specify the `mode` parameter.

In [18]:
print("Best hyperparameters from Grid Search:",
      grid_search_analysis.get_best_config("accuracy", "max", "last"))


Best hyperparameters from Grid Search: {'filters': 64, 'lr': 0.001, 'batch_size': 128, 'dropout': 0.5161961869143823}


In [8]:

! pip install "hyperopt==0.2.7"


In [1]:
pip install bayesian-optimization==1.4.3


In [9]:
from ray.tune.search.bayesopt import BayesOptSearch
import bayes_opt

In [20]:
# Bayesian search setup
import time
bayes_search_space = {
   "filters": tune.uniform(64, 256),
   "lr": tune.loguniform(1e-4, 1e-1),
   "batch_size": tune.uniform(64, 256),
   "dropout": tune.uniform(0, 1)
}

start_time = time.time()

bayesopt_search = BayesOptSearch(metric="accuracy", mode="max")
bayesian_search_analysis = tune.run(
    train_mnist,
    num_samples=8,
    config=bayes_search_space,
    search_alg=bayesopt_search,
    verbose=1,
    resources_per_trial={"cpu": 12, "gpu": 1}
)
end_time = time.time()
print("Time taken for Bayesian Search:", end_time - start_time)
print("Best hyperparameters from Bayesian Search:", bayesian_search_analysis.get_best_config("accuracy", "max", "last"))

2024-05-02 01:20:57,639	WARNING bayesopt_search.py:431 -- BayesOpt does not support specific sampling methods. The LogUniform sampler will be dropped.


+--------------------------------------------------------------------+
| Configuration for experiment     train_mnist_2024-05-02_01-20-57   |
+--------------------------------------------------------------------+
| Search algorithm                 SearchGenerator                   |
| Scheduler                        FIFOScheduler                     |
| Number of trials                 8                                 |
+--------------------------------------------------------------------+

View detailed results here: /root/ray_results/train_mnist_2024-05-02_01-20-57
To visualize your results with TensorBoard, run: `tensorboard --logdir /tmp/ray/session_2024-05-02_01-20-53_244041_927/artifacts/2024-05-02_01-20-57/train_mnist_2024-05-02_01-20-57/driver_artifacts`

Trial status: 1 PENDING
Current time: 2024-05-02 01:20:57. Total running time: 0s
Logical resource usage: 0/12 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:A100)
+-----------------------------------------------------------------

2024-05-02 01:42:37,621	WARNING experiment_state.py:205 -- Experiment state snapshotting has been triggered multiple times in the last 5.0 seconds. A snapshot is forced if `CheckpointConfig(num_to_keep)` is set, and a trial has checkpointed >= `num_to_keep` times since the last snapshot.
You may want to consider increasing the `CheckpointConfig(num_to_keep)` or decreasing the frequency of saving checkpoints.
You can suppress this error by setting the environment variable TUNE_WARN_EXCESSIVE_EXPERIMENT_CHECKPOINT_SYNC_THRESHOLD_S to a smaller value than the current threshold (5.0).
2024-05-02 01:42:37,623	INFO tune.py:1007 -- Wrote the latest version of all result files and experiment state to '/root/ray_results/train_mnist_2024-05-02_01-20-57' in 0.0063s.



Trial train_mnist_6fd9a37e completed after 5 iterations at 2024-05-02 01:42:37. Total running time: 21min 39s
+-----------------------------------------------+
| Trial train_mnist_6fd9a37e result             |
+-----------------------------------------------+
| checkpoint_dir_name                           |
| time_this_iter_s                      14.4848 |
| time_total_s                          74.5704 |
| training_iteration                          5 |
| accuracy                               0.1135 |
+-----------------------------------------------+

Trial status: 8 TERMINATED
Current time: 2024-05-02 01:42:37. Total running time: 21min 39s
Logical resource usage: 12.0/12 CPUs, 1.0/1 GPUs (0.0/1.0 accelerator_type:A100)
+-------------------------------------------------------------------------------------------------------------------------------+
| Trial name             status         filters          lr     batch_size     dropout     iter     total time (s)     accuracy |
+----

Time taken for Bayesian Search: 1300.0236146450043

Best hyperparameters from Bayesian Search: {'filters': 204.54283682778978, 'lr': 0.05990598257128396, 'batch_size': 135.91170281869358, 'dropout': 0.9507143064099162}

In [31]:
# time it
start_time = time.time()
# Hyperband setup
hyperband_scheduler = HyperBandScheduler(metric="accuracy", mode="max")
hyperband_analysis = tune.run(
    train_mnist,
    num_samples=8,
    config=search_space,
    scheduler=hyperband_scheduler,
    verbose=1,
    resources_per_trial={"cpu": 12, "gpu": 1}
)
end_time = time.time()
print("Time taken for Hyperband:", end_time - start_time)
print("Best hyperparameters from Hyperband:", hyperband_analysis.get_best_config("accuracy", "max", "last"))
ray.shutdown()

2024-05-01 01:17:07,564	INFO tune.py:762 -- Total run time: 1518.54 seconds (1518.41 seconds for the tuning loop).


Time taken for Hyperband: 1518.5629179477692
Best hyperparameters from Hyperband: {'filters': 64, 'lr': 0.001, 'batch_size': 128, 'dropout': 0.6026518587327778}


Time taken for Hyperband: 1518.5629179477692

Best hyperparameters from Hyperband: {'filters': 64, 'lr': 0.001, 'batch_size': 128, 'dropout': 0.6026518587327778}

### 2.

Time taken for Bayesian Search: 1300s = 20 minutes

Best hyperparameters from Bayesian Search: {'filters': 204.54283682778978, 'lr': 0.05990598257128396, 'batch_size': 135.91170281869358, 'dropout': 0.9507143064099162}

Time taken for Hyperband: 1518s = 25 minutes

Best hyperparameters from Hyperband: {'filters': 64, 'lr': 0.001, 'batch_size': 128, 'dropout': 0.6026518587327778}

Time taken for Grid Search: 1019s = 17 minutes

Best hyperparameters from Grid Search: {'filters': 64, 'lr': 0.001, 'batch_size': 128, 'dropout': 0.5161961869143823}





### 3.




1. **Time Efficiency**:
   - **Bayesian Search** took the longest 20 minutes.
   - **Hyperband** was slightly slower than Bayesian (about 25 minutes).
   - **Grid Search** was the fastest (about 17 minutes).

2. **Hyperparameter Values**:
   - **Bayesian Search** and **Hyperband** led to different best hyperparameters. Bayesian Search suggested much higher values for 'filters' and 'dropout' and a higher learning rate compared to Hyperband. This imply that Bayesian Search explored more varied parts of the parameter space.

   - **Grid Search** and **Hyperband** found similar hyperparameters, particularly for 'filters', 'lr', and 'batch_size'. This consistency could suggest these values are relatively stable choices within the explored grid or range.

3. **Performance Considerations**:
   - **Bayesian Search** might have explored more extreme values (e.g., very high dropout and filters), which can sometimes lead to other methods might miss due to their more conservative strategies.

   - **Hyperband** adapt the resources allocated to promising configurations. The similarity in parameters found by Hyperband and Grid Search, however, might indicate that the parameter space has a well-defined optimal region or that the range of parameters tested in these methods was somewhat limited.


## Problem 3

Learner 1 is 2.5x faster than learner 2

The staleness as follows:

$g[L_1,2]$: 0, this is the start and first of the calculation, of both learner 1 and 2

$g[L_1,2]$: 0, still there is no updated gradient from other learner during computation.

$g[L_1,3]$: 1, it fail to pick up learner 2 g[L_2, 2] which happen at 2.5s, while learner 1 is running from 2s to 3s.

$g[L_1,4]$: 0, there is no updated gradient from other learner during computation.

$g[L_2,1]$: 0, this is the start and first of the calculation, of both learner 1 and 2

$g[L_2,2]$: 2, there are 2 results from learn 1 updated during 2.5 s to 5 s: $g[L_1,3]$ and $g[L_1,4]$.

## Problem 4

### 1.

In [ ]:
!pip install torchvision

In [ ]:
'''ResNet in PyTorch.

For Pre-activation ResNet, see 'preact_resnet.py'.

Reference:
[1] Kaiming He, Xiangyu Zhang, Shaoqing Ren, Jian Sun
    Deep Residual Learning for Image Recognition. arXiv:1512.03385
'''
import torch
import torch.nn as nn
import torch.nn.functional as F


class BasicBlock(nn.Module):
    expansion = 1

    def __init__(self, in_planes, planes, stride=1):
        super(BasicBlock, self).__init__()
        self.conv1 = nn.Conv2d(
            in_planes, planes, kernel_size=3, stride=stride, padding=1, bias=False)
        self.bn1 = nn.BatchNorm2d(planes)
        self.conv2 = nn.Conv2d(planes, planes, kernel_size=3,
                               stride=1, padding=1, bias=False)
        self.bn2 = nn.BatchNorm2d(planes)

        self.shortcut = nn.Sequential()
        if stride != 1 or in_planes != self.expansion*planes:
            self.shortcut = nn.Sequential(
                nn.Conv2d(in_planes, self.expansion*planes,
                          kernel_size=1, stride=stride, bias=False),
                nn.BatchNorm2d(self.expansion*planes)
            )

    def forward(self, x):
        out = F.relu(self.bn1(self.conv1(x)))
        out = self.bn2(self.conv2(out))
        out += self.shortcut(x)
        out = F.relu(out)
        return out


class Bottleneck(nn.Module):
    expansion = 4

    def __init__(self, in_planes, planes, stride=1):
        super(Bottleneck, self).__init__()
        self.conv1 = nn.Conv2d(in_planes, planes, kernel_size=1, bias=False)
        self.bn1 = nn.BatchNorm2d(planes)
        self.conv2 = nn.Conv2d(planes, planes, kernel_size=3,
                               stride=stride, padding=1, bias=False)
        self.bn2 = nn.BatchNorm2d(planes)
        self.conv3 = nn.Conv2d(planes, self.expansion *
                               planes, kernel_size=1, bias=False)
        self.bn3 = nn.BatchNorm2d(self.expansion*planes)

        self.shortcut = nn.Sequential()
        if stride != 1 or in_planes != self.expansion*planes:
            self.shortcut = nn.Sequential(
                nn.Conv2d(in_planes, self.expansion*planes,
                          kernel_size=1, stride=stride, bias=False),
                nn.BatchNorm2d(self.expansion*planes)
            )

    def forward(self, x):
        out = F.relu(self.bn1(self.conv1(x)))
        out = F.relu(self.bn2(self.conv2(out)))
        out = self.bn3(self.conv3(out))
        out += self.shortcut(x)
        out = F.relu(out)
        return out


class ResNet(nn.Module):
    def __init__(self, block, num_blocks, num_classes=10):
        super(ResNet, self).__init__()
        self.in_planes = 64

        self.conv1 = nn.Conv2d(3, 64, kernel_size=3,
                               stride=1, padding=1, bias=False)
        self.bn1 = nn.BatchNorm2d(64)
        self.layer1 = self._make_layer(block, 64, num_blocks[0], stride=1)
        self.layer2 = self._make_layer(block, 128, num_blocks[1], stride=2)
        self.layer3 = self._make_layer(block, 256, num_blocks[2], stride=2)
        self.layer4 = self._make_layer(block, 512, num_blocks[3], stride=2)
        self.linear = nn.Linear(512*block.expansion, num_classes)

    def _make_layer(self, block, planes, num_blocks, stride):
        strides = [stride] + [1]*(num_blocks-1)
        layers = []
        for stride in strides:
            layers.append(block(self.in_planes, planes, stride))
            self.in_planes = planes * block.expansion
        return nn.Sequential(*layers)

    def forward(self, x):
        out = F.relu(self.bn1(self.conv1(x)))
        out = self.layer1(out)
        out = self.layer2(out)
        out = self.layer3(out)
        out = self.layer4(out)
        out = F.avg_pool2d(out, 4)
        out = out.view(out.size(0), -1)
        out = self.linear(out)
        return out


def ResNet18():
    return ResNet(BasicBlock, [2, 2, 2, 2])

def test():
    net = ResNet18()
    y = net(torch.randn(1, 3, 32, 32))
    print(y.size())

test()

torch.Size([1, 10])


In [ ]:

'''Train CIFAR10 with PyTorch.'''
import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
import torch.backends.cudnn as cudnn
import torchvision.transforms as transforms
from torchvision import datasets
from torch.utils.data import DataLoader
from torch.nn.parallel import DataParallel

import time
import os
import argparse

# Data
print('==> Preparing data..')
transform_train = transforms.Compose([
    transforms.RandomCrop(32, padding=4),
    transforms.RandomHorizontalFlip(),
    transforms.ToTensor(),
    transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),
])

transform_test = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),
])

trainset = torchvision.datasets.CIFAR10(
    root='./data', train=True, download=True, transform=transform_train)
trainloader = torch.utils.data.DataLoader(
    trainset, batch_size=128, shuffle=True, num_workers=2)

testset = torchvision.datasets.CIFAR10(
    root='./data', train=False, download=True, transform=transform_test)
testloader = torch.utils.data.DataLoader(
    testset, batch_size=100, shuffle=False, num_workers=2)

classes = ('plane', 'car', 'bird', 'cat', 'deer',
           'dog', 'frog', 'horse', 'ship', 'truck')

In [ ]:
net = ResNet18()
net = net.to(device)
if device == 'cuda':
    net = torch.nn.DataParallel(net)
    cudnn.benchmark = True


criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(net.parameters(), lr=args.lr,
                      momentum=0.9, weight_decay=5e-4)
scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=200)

# Training
def train(epoch, trainloader):
    # print('\nEpoch: %d' % epoch)

    net.train()
    train_loss = 0
    correct = 0
    total = 0

    for batch_idx, (inputs, targets) in enumerate(trainloader):
        inputs, targets = inputs.to(device), targets.to(device)
        optimizer.zero_grad()
        outputs = net(inputs)
        loss = criterion(outputs, targets)
        loss.backward()
        optimizer.step()

        train_loss += loss.item()
        _, predicted = outputs.max(1)
        total += targets.size(0)
        correct += predicted.eq(targets).sum().item()

        # print('Loss: %.3f | Acc: %.3f%% (%d/%d)' % (train_loss/(batch_idx+1), 100.*correct/total, correct, total))




    return end_time - start_time

batch_size = 32

while True:
    try:
        trainloader = DataLoader(trainset, batch_size, shuffle=True, num_workers=2)
        train(1, trainloader)
        start_time = time.time()  # Start timing here
        train(1, trainloader)
        end_time = time.time()  # Start timing here
        print('For batch size ', batch_size, ': ', end_time - start_time, ' sec')
        batch_size *= 4
    except RuntimeError as e:
            if 'Out of memory' in str(e):
                print(f"Out of memory with batch size {batch_size}.")



For batch size 32: 19.23 sec
For batch size 128: 13.30 sec
For batch size 512: 15.41 sec
For batch size 2048: 18.11 sec
For batch size 8192: 20.23 sec
Out of memory with batch size 32768.


When batch size increases, training time increases

### 2.

In [ ]:
net = ResNet18()
net = net.to(device)
if device == 'cuda':
    net = torch.nn.DataParallel(net)
    cudnn.benchmark = True


criterion = nn.CrossEntropyLoss()


# Training
def train(epoch, trainloader):
    # print('\nEpoch: %d' % epoch)

    net.train()
    train_loss = 0
    correct = 0
    total = 0

    for batch_idx, (inputs, targets) in enumerate(trainloader):
        inputs, targets = inputs.to(device), targets.to(device)
        optimizer.zero_grad()
        outputs = net(inputs)
        loss = criterion(outputs, targets)
        loss.backward()
        optimizer.step()

        train_loss += loss.item()
        _, predicted = outputs.max(1)
        total += targets.size(0)
        correct += predicted.eq(targets).sum().item()

        # print('Loss: %.3f | Acc: %.3f%% (%d/%d)' % (train_loss/(batch_idx+1), 100.*correct/total, correct, total))




    return end_time - start_time

batch_size = 32
gpu = [1, 2, 4]  # Configurations for 1, 2, and 4 GPUs
batch_size = 32
while True:
    try:
        for num_gpu in gpu:

            if num_gpu > 1:
                    model = DataParallel(model, device_ids=list(range(num_gpu)))
            optimizer = optim.SGD(net.parameters(), lr=args.lr,
                      momentum=0.9, weight_decay=5e-4)
            scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=200)
            trainloader = DataLoader(trainset, batchsize, shuffle=True, num_workers=2)
            train(1, trainloader) # warm up
            start_time = time.time()  # Start timing here
            train(1, trainloader)
            end_time = time.time()  # Start timing here
            print('For batch size ', batchsize, ' on ', num_gpu, ' GPUs: ', end_time - start_time, ' sec')
            batch_size *= 4
    except RuntimeError as e:
            if 'out of memory' in str(e):
                print(f"Out of memory with batch size {batch_size}.")



For batch size 32 on 1 GPUs: 19.58 sec
For batch size 32 on 2 GPUs: 49.35 sec
For batch size 32 on 4 GPUs: 68.65 sec
For batch size 128 on 1 GPUs: 19.09 sec
For batch size 128 on 2 GPUs: 23.23 sec
For batch size 128 on 4 GPUs: 22.26 sec
For batch size 512 on 1 GPUs: 12.77 sec
For batch size 512 on 2 GPUs: 8.76 sec
For batch size 512 on 4 GPUs: 7.56 sec
For batch size 2048 on 1 GPUs: 22.26 sec
For batch size 2048 on 2 GPUs: 14.56 sec
For batch size 2048 on 4 GPUs: 12.43 sec
For batch size 8192 on 1 GPUs: 23.65 sec
For batch size 8192 on 2 GPUs: 15.42 sec
For batch size 8192 on 4 GPUs: 11.34 sec
Out of memory with batch size 32768 on 1 GPUs.


| Batch-size per GPU |           | Batch-size 32 |            | Batch-size 128 |           | Batch-size 512 |            | Batch-size 2048 |           | Batch-size 8192 |           |
|--------------------|-----------|---------------|------------|----------------|-----------|----------------|------------|-----------------|-----------|-----------------|-----------|
|                    | Time(sec) | Speedup       | Time(sec)  | Speedup        | Time(sec) | Speedup        | Time(sec)  | Speedup         | Time(sec) | Speedup         |
| **1-GPU**          | 19.58     | 1             | 19.09      | 1              | 12.77     | 1              | 22.26      | 1               | 23.65     | 1               |
| **2-GPU**          | 49.35     | 0.397         | 23.23      | 0.822          | 8.76      | 1.457          | 14.56      | 1.529           | 15.42     | 1.534           |
| **4-GPU**          | 68.65     | 0.285         | 22.26      | 0.858          | 7.56      | 1.689          | 12.43      | 1.791           | 11.34     | 2.085           |


we are measuring strong-scaling. Strong-scaling focuses on how the problem size remains constant while the number of processing elements (in this case, GPUs) is increased. The primary goal of strong-scaling is to reduce the time to solution by using more computational resources.

If Weak-Scaling Was Used:

Weak-scaling measures the scalability of the system by increasing the size of the problem proportionally as more processing units are added, aiming to keep the workload per processing unit constant.

the time for computation might not degrade as it does currently with smaller batch sizes on more GPUs. This would likely result in a more linear scaling, which means the speedup would be closer to the ideal since each GPU would have enough work to perform efficiently. So, **we would expect to see less dramatic increases in time or even reductions in time with more GPUs for smaller batch sizes, and possibly even better performance improvements for larger batch sizes.**

for smaller batch sizes, strong-scaling is not very effective due to overheads outweighing benefits. For larger batch sizes, strong-scaling shows better performance improvements as the computational load justifies the communication and synchronization overhead.

### 3.

Compute time for a multi-GPU setup is estimated by dividing the 1 GPU training time by the number of GPUs. This assumes perfect scaling from 1 GPU to multiple GPUs, i.e., zero communication overhead.

Communication time is the excess time taken for training on multiple GPUs over the calculated compute time. So, Communication Time=Total Time on N GPUs−Compute Time on N GPUs

| Batch-size per GPU |              | Batch-size 32 |                | Batch-size 128 |               | Batch-size 512 |                | Batch-size 2048 |               | Batch-size 8192 |               |
|--------------------|--------------|---------------|----------------|----------------|---------------|----------------|----------------|-----------------|---------------|-----------------|---------------|
|                    | Compute(sec) | Comm(sec)     | Compute(sec)   | Comm(sec)      | Compute(sec)  | Comm(sec)      | Compute(sec)   | Comm(sec)      | Compute(sec)  | Comm(sec)      |
| **2-GPU**          | 9.79         | 39.56         | 9.545          | 13.685         | 6.385         | 2.375          | 11.13          | 3.43            | 11.825        | 3.595          |
| **4-GPU**          | 4.895        | 63.755        | 4.7725         | 17.4875        | 3.1925        | 4.3675         | 5.565          | 6.865           | 5.9125        | 5.4275         |


### 4.

Formula:

$T_{all_reduce} = 2(P — 1)* S/B$

where P is the number of GPUs participating in the all-reduce.

S is the total size of the data to be reduced across all GPUs, usually in bytes. This is model param for ResNet, which is 11689512

B is the communication bandwidth per link between GPUs, usually in bytes per second.

each GPU sends and receives data of size S to/from its neighbors and does this for P−1 rounds.

Formula for Bandwidth Utilization:

U=  Actual Data Transferred / time
​
Actual Data Transferred is 2×(P−1)×S (since each GPU sends and receives S byte P−1 times).

So, for 2 GPU, the ADT = 2*(2 — 1)*11689512 = 11689512 = 0.045GB

for 4 GPU, the ADT = 2*(4 — 1)*11689512 = 17534268 = 0.065GB



| Batch-size-per-GPU | Bandwidth Utilization(GB/s)  | Bandwidth Utilization(GB/s) | Bandwidth Utilization(GB/s) | Bandwidth Utilization(GB/s) | Bandwidth Utilization(GB/s) |
|--------------------|------------------------------|------------------------------|------------------------------|-----------------------------|------------------------------|
|                    | Batch-size 32                | Batch-size 128               | Batch-size 512               | Batch-size 2048             | Batch-size 8192              |
| **2-GPU**          | 0.000912                     | 0.001937                     | 0.005137                     | 0.003089                    | 0.002918                     |
| **4-GPU**          | 0.000947                     | 0.00292                      | 0.008598                     | 0.005231                    | 0.005731                     |



As the batch size increases, bandwidth utilization becomes more efficient due to the larger amount of data being transferred relative to the communication overhead.

## Problem 5

### 1.

In [ ]:
!git clone https://github.com/qfgaohao/pytorch-ssd.git

Cloning into 'pytorch-ssd'...
remote: Enumerating objects: 819, done.
remote: Counting objects: 100% (437/437), done.
remote: Compressing objects: 100% (32/32), done.
remote: Total 819 (delta 411), reused 405 (delta 405), pack-reused 382
Receiving objects: 100% (819/819), 1.05 MiB | 3.51 MiB/s, done.
Resolving deltas: 100% (552/552), done.


In [ ]:
pip install opencv-python tqdm

Note: you may need to restart the kernel to use updated packages.


In [ ]:
!python pytorch-ssd/eval_ssd.py --net mb1-ssd  --dataset data/VOCdevkit/VOC2007 --trained_model pytorch-ssd/models/mobilenet-v1-ssd-mp-0_675.pth --label_file pytorch-ssd/models/voc-model-labels.txt

The result is as follow, (copy and paste from hpc from above command)

Average Precision Per-class:

aeroplane: 0.6843271224059599

bicycle: 0.7911140237662206

bird: 0.6171819168583986

boat: 0.5612220055063379

bottle: 0.3485216621466003

bus: 0.7677814849265677

car: 0.7280986468467315

cat: 0.8369208203985581

chair: 0.5169138632991064

cow: 0.6238697603075337

diningtable: 0.7062172972736019

dog: 0.7872868219961326

horse: 0.819446325939355

motorbike: 0.7918539457195842

person: 0.702363739134837

pottedplant: 0.39852951468542563

sheep: 0.6066678298227772

sofa: 0.7573083661544429

train: 0.8262441264750008

tvmonitor: 0.6461898726506375

Average Precision Across All Classes: 0.6759029573156905

### 2.

In [ ]:
pip install boto3

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 139.3/139.3 kB 4.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.2/12.2 MB 35.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 82.2/82.2 kB 7.2 MB/s eta 0:00:00


In [ ]:
!python pytorch-ssd/open_images_downloader.py --root data/open_images --class_names "Handgun,Shotgun" --num_workers 20

2024-04-28 19:50:21,730 - root - Download https://storage.googleapis.com/openimages/2018_04/class-descriptions-boxable.csv.
2024-04-28 19:50:21,842 - root - Download https://storage.googleapis.com/openimages/2018_04/train/train-annotations-bbox.csv.
2024-04-28 19:50:26,623 - root - Read annotation file data/open_images/train-annotations-bbox.csv
2024-04-28 19:50:43,293 - root - train bounding boxes size: 1307
2024-04-28 19:50:43,293 - root - Approximate Image Stats: 
2024-04-28 19:50:43,297 - root - Handgun: 561/990 = 0.57.
2024-04-28 19:50:43,297 - root - Shotgun: 429/990 = 0.43.
2024-04-28 19:50:43,297 - root - Label distribution: 
2024-04-28 19:50:43,297 - root - Handgun: 727/1307 = 0.56.
2024-04-28 19:50:43,297 - root - Shotgun: 580/1307 = 0.44.
2024-04-28 19:50:43,297 - root - Shuffle dataset.
2024-04-28 19:50:43,297 - root - Save train data to data/open_images/sub-train-annotations-bbox.csv.
2024-04-28 19:50:43,307 - root - Download https://storage.googleapis.com/openimages/2018_

Download successfully, I choose Handguns and Shotguns.

In [ ]:
%cd pytorch-ssd

/content/pytorch-ssd


In [ ]:
!python train_ssd.py --dataset_type open_images --datasets ../data/open_images --net mb1-ssd --pretrained_ssd models/mobilenet-v1-ssd-mp-0_675.pth --scheduler cosine --lr 0.01 --t_max 100 --validation_epochs 5 --num_epochs 100 --base_net_lr 0.001  --batch_size 5

2024-04-28 19:52:05,584 - root - INFO - Use Cuda.
2024-04-28 19:52:05,584 - root - INFO - Namespace(dataset_type='open_images', datasets=['../data/open_images'], validation_dataset=None, balance_data=False, net='mb1-ssd', freeze_base_net=False, freeze_net=False, mb2_width_mult=1.0, lr=0.01, momentum=0.9, weight_decay=0.0005, gamma=0.1, base_net_lr=0.001, extra_layers_lr=None, base_net=None, pretrained_ssd='models/mobilenet-v1-ssd-mp-0_675.pth', resume=None, scheduler='cosine', milestones='80,100', t_max=100.0, batch_size=5, num_epochs=100, num_workers=4, validation_epochs=5, debug_steps=100, use_cuda=True, checkpoint_folder='models/')
2024-04-28 19:52:05,618 - root - INFO - Prepare training datasets.
2024-04-28 19:52:06,164 - root - INFO - Dataset Summary:Number of Images: 961
Minimum Number of Images for a Class: -1
Label Distribution:
	Handgun: 727
	Shotgun: 580
2024-04-28 19:52:06,165 - root - INFO - Stored labels into file models/open-images-model-labels.txt.
2024-04-28 19:52:06,16

In [ ]:
!python eval_ssd.py --dataset_type open_images --net mb1-ssd --dataset ../data/open_images --trained_model models/mb1-ssd-Epoch-99-Loss-2.798509557247162.pth --label_file models/open-images-model-labels.txt

/content/pytorch-ssd/eval_ssd.py:69: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  all_gt_boxes[class_index][image_id] = torch.tensor(all_gt_boxes[class_index][image_id])
It took 0.04636216163635254 sec to load the model.
process image 0
Load Image: 0.019361 sec.
Inference time:  0.6022565364837646
Prediction: 0.611756 sec.
process image 1
Load Image: 0.009265 sec.
Inference time:  0.005998373031616211
Prediction: 0.016124 sec.
process image 2
Load Image: 0.007794 sec.
Inference time:  0.00577545166015625
Prediction: 0.014228 sec.
process image 3
Load Image: 0.010273 sec.
Inference time:  0.006129026412963867
Prediction: 0.010442 sec.
process image 4
Load Image: 0.014009 sec.
Inference time:  0.0057332515716552734
Prediction: 0.011431 sec.
process image 5
Load Image: 0.008108 sec.
Inference time:  0.006911754608154297
Prediction: 0.0199

In [ ]:
!python eval_ssd.py --dataset_type open_images --net mb1-ssd --dataset ../data/open_images --trained_model models/mb1-ssd-Epoch-0-Loss-3.7128826236724852.pth --label_file models/open-images-model-labels.txt

/content/pytorch-ssd/eval_ssd.py:69: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  all_gt_boxes[class_index][image_id] = torch.tensor(all_gt_boxes[class_index][image_id])
It took 0.04904580116271973 sec to load the model.
process image 0
Load Image: 0.019377 sec.
Inference time:  0.5990192890167236
Prediction: 0.621946 sec.
process image 1
Load Image: 0.009439 sec.
Inference time:  0.0062274932861328125
Prediction: 0.025332 sec.
process image 2
Load Image: 0.007796 sec.
Inference time:  0.005885124206542969
Prediction: 0.026655 sec.
process image 3
Load Image: 0.010315 sec.
Inference time:  0.006087541580200195
Prediction: 0.030978 sec.
process image 4
Load Image: 0.014173 sec.
Inference time:  0.006224393844604492
Prediction: 0.025230 sec.
process image 5
Load Image: 0.008260 sec.
Inference time:  0.005950450897216797
Prediction: 0.026

### 3.

In [ ]:
!pip install onnx onnxruntime netron

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.5/16.5 MB 23.7 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.8/14.8 MB 25.6 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 24.4 MB/s eta 0:00:00a 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 404.0/404.0 kB 17.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.0/46.0 kB 5.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.8/86.8 kB 12.5 MB/s eta 0:00:00


In [ ]:
%cd pytorch-ssd

/Users/zhongyihao/Downloads/DS301 Advance ML/deep-learning/code/pytorch-ssd


In [ ]:
# Some standard imports
import numpy as np

from torch import nn
import torch.utils.model_zoo as model_zoo
import torch.onnx

from vision.ssd.mobilenetv1_ssd import create_mobilenetv1_ssd
import torch.onnx
import onnx

device = 'cuda' if torch.cuda.is_available() else 'cpu'


# Parameters

model_path = 'models/mb1-ssd-Epoch-99-Loss-2.798509557247162.pth'
label_path = 'models/open-images-model-labels.txt'

class_names = [name.strip() for name in open(label_path).readlines()]


net = create_mobilenetv1_ssd(len(class_names), is_test=True)
net.load(model_path)
net.eval()

net.to(device)

model_onnx_path = f"models/mb1-ssd-10.onnx"

x = torch.randn(1, 3, 300, 300, requires_grad=True).to(device)

torch.onnx.export(net,
                  x,
                  model_onnx_path,
                  verbose=False,
                  export_params=True,
                  output_names=['scores', 'boxes'],
                  input_names = ['input'],   # the model's input names
                  dynamic_axes={'input' : {0 : 'batch_size'},    # variable length axes
                                'output' : {0 : 'batch_size'}})

# Load the ONNX model and print a message
model = onnx.load(model_onnx_path)
print(f"File saved: {model_onnx_path}.")


/Users/zhongyihao/anaconda3/envs/torch_nightly_env/lib/python3.10/site-packages/torch/onnx/utils.py:2029: UserWarning: Provided key output for dynamic axes is not a valid input/output name
  warnings.warn(


================ Diagnostic Run torch.onnx.export version 2.0.0 ================
verbose: False, log level: Level.ERROR
======================= 0 NONE 0 NOTE 0 WARNING 0 ERROR ========================

File saved: models/mb1-ssd-10.onnx.


### 4.

In [ ]:

onnx_model = onnx.load("models/mb1-ssd-10.onnx")

# Check
onnx.checker.check_model(onnx_model)
print("Check")

Check


### 5.

In [ ]:
import onnx
import onnxruntime

onnx_model = onnx.load(model_onnx_path)
onnx.checker.check_model(onnx_model)
print(onnx.helper.printable_graph(onnx_model.graph))
ort_session = onnxruntime.InferenceSession(model_onnx_path)

def to_numpy(tensor):
    return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()

# ONNX Runtime output prediction
ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(x)}
ort_outs = ort_session.run(None, ort_inputs)

print(ort_outs[0].shape)
print(ort_outs[1].shape)

graph torch_jit (
  %input[FLOAT, batch_sizex3x300x300]
) initializers (
  %extras.0.0.weight[FLOAT, 256x1024x1x1]
  %extras.0.0.bias[FLOAT, 256]
  %extras.0.2.weight[FLOAT, 512x256x3x3]
  %extras.0.2.bias[FLOAT, 512]
  %extras.1.0.weight[FLOAT, 128x512x1x1]
  %extras.1.0.bias[FLOAT, 128]
  %extras.1.2.weight[FLOAT, 256x128x3x3]
  %extras.1.2.bias[FLOAT, 256]
  %extras.2.0.weight[FLOAT, 128x256x1x1]
  %extras.2.0.bias[FLOAT, 128]
  %extras.2.2.weight[FLOAT, 256x128x3x3]
  %extras.2.2.bias[FLOAT, 256]
  %extras.3.0.weight[FLOAT, 128x256x1x1]
  %extras.3.0.bias[FLOAT, 128]
  %extras.3.2.weight[FLOAT, 256x128x3x3]
  %extras.3.2.bias[FLOAT, 256]
  %classification_headers.0.weight[FLOAT, 18x512x3x3]
  %classification_headers.0.bias[FLOAT, 18]
  %classification_headers.1.weight[FLOAT, 18x1024x3x3]
  %classification_headers.1.bias[FLOAT, 18]
  %classification_headers.2.weight[FLOAT, 18x512x3x3]
  %classification_headers.2.bias[FLOAT, 18]
  %classification_headers.3.weight[FLOAT, 18x256x3x3]
 

### 6.

In [ ]:
import onnx
import onnxruntime

onnx_model = onnx.load(model_onnx_path)
onnx.checker.check_model(onnx_model)
print(onnx.helper.printable_graph(onnx_model.graph))
ort_session = onnxruntime.InferenceSession(model_onnx_path)

def to_numpy(tensor):
    return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()

# ONNX Runtime output prediction
ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(x)}
ort_outs = ort_session.run(None, ort_inputs)


# PyTorch output prediction
torch_out = net(x)
torch_out_np = [to_numpy(o) for o in torch_out]

print("output shape of ort", ort_outs[0].shape)
print("output shape of torch", torch_out_np[0].shape)

# compare ONNX Runtime and PyTorch results
np.testing.assert_allclose(torch_out_np[0], ort_outs[0], rtol=1e-02, atol=1e-05)

print("Exported model has been tested with ONNXRuntime, and the result looks good!")

graph torch_jit (
  %input[FLOAT, batch_sizex3x300x300]
) initializers (
  %extras.0.0.weight[FLOAT, 256x1024x1x1]
  %extras.0.0.bias[FLOAT, 256]
  %extras.0.2.weight[FLOAT, 512x256x3x3]
  %extras.0.2.bias[FLOAT, 512]
  %extras.1.0.weight[FLOAT, 128x512x1x1]
  %extras.1.0.bias[FLOAT, 128]
  %extras.1.2.weight[FLOAT, 256x128x3x3]
  %extras.1.2.bias[FLOAT, 256]
  %extras.2.0.weight[FLOAT, 128x256x1x1]
  %extras.2.0.bias[FLOAT, 128]
  %extras.2.2.weight[FLOAT, 256x128x3x3]
  %extras.2.2.bias[FLOAT, 256]
  %extras.3.0.weight[FLOAT, 128x256x1x1]
  %extras.3.0.bias[FLOAT, 128]
  %extras.3.2.weight[FLOAT, 256x128x3x3]
  %extras.3.2.bias[FLOAT, 256]
  %classification_headers.0.weight[FLOAT, 18x512x3x3]
  %classification_headers.0.bias[FLOAT, 18]
  %classification_headers.1.weight[FLOAT, 18x1024x3x3]
  %classification_headers.1.bias[FLOAT, 18]
  %classification_headers.2.weight[FLOAT, 18x512x3x3]
  %classification_headers.2.bias[FLOAT, 18]
  %classification_headers.3.weight[FLOAT, 18x256x3x3]
 

output shape of ort (1, 3000, 3)

output shape of torch (1, 3000, 3)

Exported model has been tested with ONNXRuntime, and the result looks good!

The precision used for matching is determined by the rtol and atol parameters:

`rtol=1e-02`: This sets the relative tolerance to 1e-02, which is equivalent to 0.01 or 1%. It means that the relative difference between the corresponding elements of the two arrays should be within 1% of each other.

`atol=1e-05`: This sets the absolute tolerance to 1e-05, which is equivalent to 0.00001. It means that the absolute difference between the corresponding elements of the two arrays should be within 0.00001.

It uses `abs(a - b) <= atol + rtol * abs(b)` to calculate the difference between the two object

### 7.

In [ ]:
pwd

'/Users/zhongyihao/Downloads/DS301 Advance ML/deep-learning/code/pytorch-ssd'

In [ ]:
from PIL import Image
import torchvision.transforms as transforms
import onnx
import onnxruntime

img_list = ['../data/open_images/train/0002f32727ed756e.jpg',
              '../data/open_images/train/000305928f7b43da.jpg']

ort_session = onnxruntime.InferenceSession("models/mb1-ssd-10.onnx")

# Preprocess and normalize the image
def pre(img):
    img_data = np.array(img)
    img_data = np.transpose(img_data, [2, 0, 1])
    img_data = np.expand_dims(img_data, 0)
    mean_vec = np.array([0.485, 0.456, 0.406])
    stddev_vec = np.array([0.229, 0.224, 0.225])
    norm_img_data = np.zeros(img_data.shape).astype('float32')
    for i in range(img_data.shape[1]):
        norm_img_data[:,i,:,:] = (img_data[:,i,:,:]/255 - mean_vec[i]) / stddev_vec[i]

for img_path in img_list:
    img = Image.open(img_path)
    img.show()

    resize = transforms.Resize([300, 300])
    img = resize(img)

    img_ycbcr = img.convert('YCbCr')
    img_y, img_cb, img_cr = img_ycbcr.split()

    to_tensor = transforms.ToTensor()
    img_y = to_tensor(img_y)
    img_y.unsqueeze_(0)

    img_3ch = img_y.repeat(1, 3, 1, 1)
    ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(img_3ch)}
    ort_outs = ort_session.run(None, ort_inputs)
    print(ort_outs[0])
    img_out_y = np.asarray(ort_outs[0])
    label = np.argmax(ort_outs[0], axis=-1)
    print(f"# of handguns: {(label == 1).sum ()}, # of shotguns: {(label == 2).sum ()}")

[[[0.99311006 0.00260478 0.00428523]
  [0.9951139  0.00106808 0.003818  ]
  [0.9907872  0.00311253 0.00610025]
  ...
  [0.9480198  0.00858595 0.04339429]
  [0.9679258  0.00882698 0.02324719]
  [0.9450335  0.00878914 0.04617732]]]
# of handguns: 0, # of shotguns: 0
[[[0.9942662  0.00391159 0.00182216]
  [0.9971625  0.00150818 0.00132924]
  [0.9937894  0.00386346 0.00234707]
  ...
  [0.8499492  0.10107069 0.0489802 ]
  [0.9336622  0.05180439 0.01453347]
  [0.8160332  0.11915151 0.06481532]]]
# of handguns: 0, # of shotguns: 0


In [ ]:
img_out_y = Image.fromarray(np.uint8((img_out_y[0] * 255.0).clip(0, 255)[0]), mode='L')

# get the output image follow post-processing step from PyTorch implementation
final_img = Image.merge(
    "YCbCr", [
        img_out_y,
        img_cb.resize(img_out_y.size, Image.BICUBIC),
        img_cr.resize(img_out_y.size, Image.BICUBIC),
    ]).convert("RGB")

# Save the image\
final_img.save("cat_superres_with_ort.jpg")

### 8.

In [ ]:
## Display image with bounding boxes and appropriate class

# Parse the list of class labels
classes = [line.rstrip('\n') for line in open('assets/coco_classes.txt')]

# Plot the bounding boxes on the image
plt.figure()
fig, ax = plt.subplots(1, figsize=(12,9))
ax.imshow(img)

resized_width = 300  # we resized the original image, remember ?
resized_height = 300
num_boxes = 1 # we limit displaying to just 10 boxes to avoid clogging the result image with boxes
               # The results are already sorted based on box confidences, so we just pick top N boxes without sorting

for c in range(num_boxes):
    base_index = c * 4
    y1, x1, y2, x2 = bboxes[base_index] * resized_height, bboxes[base_index + 1] * resized_width, bboxes[base_index + 2] * resized_height, bboxes[base_index + 3] * resized_width
    color = 'blue'
    box_h = (y2 - y1)
    box_w = (x2 - x1)
    bbox = patches.Rectangle((y1, x1), box_h, box_w, linewidth=2, edgecolor=color, facecolor='none')
    ax.add_patch(bbox)
    plt.text(y1, x1, s=classes[labels[c] - 1], color='white', verticalalignment='top', bbox={'color': color, 'pad': 0})
plt.axis('off')

# Save image
plt.savefig("output/ssd_result.jpg", bbox_inches='tight', pad_inches=0.0)
plt.show()

## Problem 6

1

In episodic tasks, the agent-environment interaction is broken into discrete episodes that terminate after a certain number of time steps or when the agent reaches a terminal state. Examples include games like chess or Go where there is a clear endpoint.

In continuous tasks, the interaction continues without limit. The agent continually interacts with the environment without explicit episodes or termination. An example can an agent controlling a power plant or traffic lights that need to continous operation.



2.

Exploration refers to the agent trying new actions to discover strategies that lead to high rewards in the long run. Exploitation means the agent leveraging its current knowledge to choose the best known action to maximize immediate reward.

Actors employ $ε$-greedy policies to balance exploration and exploitation. With probability $ε$, the agent selects a random action (exploration). With probability 1-$ε$, it selects the action with the highest estimated value (exploitation).

The $ε$ value should follow a schedule, starting high to encourage exploration early in training, then decreasing over time to shift priority to exploitation as the agent gains knowledge. A fixed $ε$ would not allow this strategic shift.

Higher $ε$ values bias the agent to explore, while lower values prioritize exploiting current knowledge. Gradually moving from high to low $ε$ during training allows the agent to discover high-value strategies early on, then increasingly leverage that knowledge over time to maximize performance.


3.

Deep Q-Learning uses a deep neural network (Q-network) to approximate the action-value function Q(s,a), whereas Q-learning uses a tabular representation. This allows Deep Q-Learning to handle high-dimensional state spaces. Also, Deep Q-Learning uses a separate target network Q' to calculate the target Q-values used in the loss function. The target network's weights are periodically copied from the main Q-network. This stabilizes learning by reducing correlations between the target and current Q-values. Q-learning uses the same Q-function for action selection and Q-value updates.

The Deep Q-Learning algorithm in Mnih et al. (2013) page 5 proceeds as follows:

Initialize the replay memory D to capacity N. This is where experienced transitions will be stored for later training.
Initialize the action-value function Q with random weights. This is the main Q-network that will be learned.

For each episode:

- Initialize the starting state s_1
- For each time step t in the episode:

   - With probability ε, select a random action a_t, otherwise select a_t = argmax_a Q(s_t,a) (ε-greedy policy)

   - Execute a_t in the emulator and observe the reward r_t and next state s_t+1

   - Store the transition (s_t, a_t, r_t+1, s_t+1) in the replay memory D

   - Sample a random minibatch of transitions from D
   - For each sampled transition (s_j, a_j, r_j+1, s_j+1):

      - If s_j+1 is terminal (episode ended), y_j = r_j

      - Otherwise, set y_j = r_j + γ * max_a' Q'(s_j+1, a') using the target network


   - Perform a gradient descent step on (y_j - Q(s_j, a_j))^2 w.r.t. the Q-network weights

- After every C steps, copy the Q-network weights to the target network Q'





4.

The target Q-network is a important component in Deep Q-Learning that helps stabilize the learning process and improve convergence. In Q-learning, the same Q-function is used to select actions and calculate target Q values for updates, which can lead to instability due to the constantly shifting targets. Also, by doing a separate target network with fixed parameters that are periodically copied from the main Q-network, the target Q values remain more consistent, providing a stable objective for the main Q-network to learn towards. This is important when combined with experience replay, as the fixed target ensures that the Q-values of sampled past transitions remain independent of recent updates to the main Q-network, this kind of breaking correlations and allowing for more efficient learning.

5.

Experience replay improves the efficiency and stability of the learning process. By storing agent's experiences in a replay buffer and randomly sampling mini-batches from it during training, experience replay breaks the correlations between sequential experiences, allows the agent to learn from past experiences multiple times, and reduces the impact of rare or noisy experiences. This leads to more efficient use of the collected data, faster convergence to the optimal Q-function, and improved stability of the learning process. Also, experience replay can do better exploration by enabling the agent to learn from experiences collected under different policies, discovering some better strategies that might not be immediately apparent with current policy.

6.

Prioritized experience replay is an extension of the standard experience replay technique that aims to further improve the efficiency of learning by prioritizing the experiences from which the agent can learn the most. Each experience is assigned a priority value that determines the probability of it being sampled during training. The priority is based on the magnitude of the temporal-difference (TD) error, which measures the difference between the predicted Q-value and the target Q-value for a given experience.

The priority $p_i$ for experience $i$ is calculated as:

$p_i = |\delta_i| + \epsilon$

where $|\delta_i|$ is the absolute value of the TD error for experience $i$, and $\epsilon$ is a small positive constant to ensure that all experiences have a non-zero probability of being sampled.

The TD error $\delta_i$ is computed as:

$\delta_i = r + \gamma \max_{a'} Q(s', a') - Q(s, a)$

where $r$ is the immediate reward, $\gamma$ is the discount factor, $s'$ is the next state, and $a'$ is the action taken in the next state.

Experiences with higher TD errors are more informative because they indicate a larger discrepancy between the predicted and target Q-values. By assigning higher priorities to these experiences, prioritized experience replay encourages the agent to focus on the most valuable experiences during training, accelerating the learning process.

To sample experiences from the replay buffer, the priorities are first converted into probabilities using the softmax function:

$P(i) = \frac{p_i^{\alpha}}{\sum_k p_k^{\alpha}}$

where $\alpha$ is a hyperparameter that determines the degree of prioritization. A higher value of $\alpha$ results in stronger prioritization, while $\alpha = 0$ corresponds to uniform sampling.

During training, experiences are sampled from the replay buffer according to their probabilities $P(i)$. The agent then updates the Q-network using the sampled experiences, giving more weight to the experiences with higher priorities. The priorities of the sampled experiences are then updated based on the new TD errors, ensuring that the agent continues to prioritize the most informative experiences throughout the learning process.

7.

Similarities:

Distributed architecture: Both GORILA and Ape-X employ a distributed architecture that separates the learning process into multiple components, for parallel processing and faster learning.

Experience replay: Both architectures use experience replay to store and sample past experiences for training. It helps break correlations between sequential experiences, improves data efficiency.

Target network: GORILA and Ape-X use a target network to stabilize the learning process. The target network's parameters are periodically updated with the main Q-network's parameters to provide a more stable learning target.

Differences:

Prioritization: Ape-X employs prioritized experience replay, where experiences are prioritized based on their temporal-difference (TD) error. This allows the agent to focus on the most informative experiences during training. GORILA uses standard experience replay without prioritization.

Architecture components: GORILA consists of actors, learners, and a parameter server. Actors generate experiences, learners compute gradients and update the model parameters, and the parameter server maintains the global Q-network parameters. In contrast, Ape-X has actors, a learner, and a replay buffer. Actors generate experiences and store them in the replay buffer, while the learner samples from the replay buffer and updates the model parameters.


## Problem 7

1.

All four platform support Tensorflow (commonly version 2.XX), Pytorch (commonly version 2.XX), sci-learn (up to v1.1), Keras (via torch 2.0) and XGBoost (up to v1.6). All these packages have the option to downgrade per needed in all four platforms, providing flexibility and customization for client need.

For IBM Watson, it also support Caffe and SPSS. It also offers AutoAI, which automates model selection and hyperparameter tuning.

For Google Cloud, it centralized all tools in Vertex AI for both tranditional AI (prediction and classification) and LLMs. It provides pre-built Docker images for these frameworks, making it easy to get started with deep learning

For Amazon AWS, its machine learning platform is called SageMaker, allowing you to create, train, and deploy models using popular frameworks like TensorFlow, PyTorch, and MXNet. It also support AWS Lambda to fasten and streamline the machine learning process via these ml platform.



2.

Amazon AWS: Amazon SageMaker provides GPU instances for training deep learning models. various GPU types, including NVIDIA Tesla V100, P3, and G4 instances.

Microsoft Azure: Azure Machine Learning offers GPU support with different types of virtual machines, such as NC-series (NVIDIA Tesla K80), ND-series (NVIDIA Tesla P40), and NV-series (NVIDIA Tesla V100, A100, etc).

Google Cloud AI Platform: Google provides GPU support through Compute Engine instances. You can select from NVIDIA Tesla K80, P4, L4, and V100, A100 GPUs.

IBM Watson Machine Learning: NV-series (NVIDIA Tesla V100, etc)



3.

Amazon AWS:

Amazon SageMaker: SageMaker provides tools for managing the entire model lifecycle. It supports versioning, model deployment, and monitoring. You can create custom pipelines for data preprocessing, model training, and deployment.

Microsoft Azure:

Azure Machine Learning: Azure ML Studio offers features for model versioning, deployment, and monitoring. It integrates with Azure DevOps for continuous integration and deployment (CI/CD) workflows. You can create reusable pipelines using Azure ML Pipelines.

Google Cloud AI Platform:

Google AI Platform: Google provides tools for model versioning, deployment, and monitoring. You can use AI Platform Pipelines to create end-to-end workflows. Additionally, Kubeflow Pipelines can be integrated for more complex scenarios.

IBM Watson Machine Learning:

IBM’s platform supports model versioning, deployment, and monitoring. It offers model catalogs, deployment spaces, and continuous learning capabilities. You can manage models using Watson Studio and Watson Machine Learning service.

4.

Amazon AWS:

Amazon SageMaker:

SageMaker provides monitoring capabilities for deployed models. It track metrics, set up alerts, and visualize performance using Amazon CloudWatch. It also supports distributed monitoring for multi-node training jobs.

Microsoft Azure:

Azure Machine Learning:

Azure ML provides monitoring through Azure Monitor. It support collect telemetry data, monitor resource usage (CPU, GPU, memory), and set up alerts. Azure Application Insights can be used for application-level monitoring.

Google Cloud AI Platform:

Google AI Platform: Google offers monitoring features through Cloud Monitoring. It can track resource utilization, set up alerts, and visualize metrics. Stackdriver can be used for logging and debugging.

IBM Watson Machine Learning:

IBM’s platform provides monitoring for deployed models. You can track performance metrics, resource usage, and model drift. Watson OpenScale offers additional capabilities for fairness, explainability, and bias detection.


5.

Amazon AWS (Amazon SageMaker):

Accuracy Visualization: Amazon SageMaker provides built-in metrics visualization during training. Someone can monitor accuracy using Amazon CloudWatch metrics or custom metrics defined in the training script. SageMaker also supports real-time visualization of training progress.

Throughput Visualization: While throughput metrics are not directly visualized, we can analyze training logs and resource utilization to estimate throughput. SageMaker offers detailed logs for debugging and performance analysis.

Microsoft Azure (Azure Machine Learning):

Accuracy Visualization: Azure ML Studio allows you to visualize accuracy metrics during training. You are able to track accuracy over epochs and visualize it using charts. Additionally, you can log custom metrics for more detailed analysis.

Throughput Visualization: Azure Monitor provides insights into resource usage, including GPU and CPU utilization. By monitoring these metrics during training, you can indirectly see throughput.

Google Cloud AI Platform:

Accuracy Visualization: Google AI Platform supports TensorFlow, which includes TensorBoard for visualizing accuracy and other metrics during training. Track accuracy, loss, and other relevant metrics.

Throughput Visualization: Similar to Azure, Google Cloud Monitoring provides resource utilization metrics. By analyzing GPU and CPU usage, you can see throughput.

IBM Watson Machine Learning:

Accuracy Visualization: IBM Watson Studio integrates with Jupyter notebooks, allowing you to visualize accuracy metrics during training. Create custom plots and track model performance.

Throughput Visualization: While Watson Studio doesn’t offer direct throughput visualization, we can analyze training logs and resource usage to see throughput patterns.

6.


Suppose we want to train a deep learning model for image classification using the popular ResNet-50 architecture. Our training job description will include the following common fields for all 4 platforms:

**Job Name:** A unique identifier for the training job.

**Experiment Name:** Organizes the job within the ML platform (e.g., Azure ML Studio’s experiments).

**Description:** Optional text describing the purpose or details of the job.

**Timeout:** Specifies the maximum allowed duration for the training job.

**Tags:** Optional tags for organization and metadata.

See below yaml, please noted that sometimes Azure allows you to use JSON format training script. The specific format depends on the platform, but the data fields and configuration are specified as above.

https://learn.microsoft.com/en-us/azure/machine-learning/how-to-train-with-ui?view=azureml-api-2

https://dataplatform.cloud.ibm.com/docs/content/wsj/analyze-data/ml_dlaas_working_with_new_models.html?context=cpdaas

https://resources.workable.com/machine-learning-engineer-job-description

https://cloud.google.com/vertex-ai/docs/tabular-data/classification-regression/prepare-data

````yaml
job.yml


job_name: resnet50-training-job
experiment_name: image-classification-experiment
description: Training ResNet-50 for image classification
timeout_hours: 8
compute_instance: cloud_platform_east_v1234
compute_cluster: username_1221
tags:
  - deep-learning
  - image-classification
````

In all platforms, we specify the job name, experiment name, description, timeout, and optional tags. The actual training script, input data, and other specifics would be added based on the specific platform’s requirements. Each platform may have additional fields or variations like data pipeline config, data strorage config, logging, post-training monitoring etc.

Common structure is as follow

```tree
.
├── job.yml (training job config file)
├── data
└── src
    └── main.py
```